packages/yew/src/functional/hooks/use_callback.rs

```diff
@@ -0,0 +1,97 @@
+use std::cell::RefCell;
+
+use crate::callback::Callback;
+use crate::functional::{hook, use_state};
+
+/// Get a immutable reference to a memoized `Callback`.
+///
+/// Memoization means it will only get recreated when provided dependencies update/change.
+/// This is useful when passing callbacks to optimized child components that rely on
+/// PartialEq to prevent unnecessary renders.
+///
+/// # Example
+///
+/// ```rust
+/// # use yew::prelude::*;
+/// #
+/// #[derive(Properties, PartialEq)]
+/// pub struct Props {
+///     pub callback: Callback<String, String>,
+/// }
+///
+/// #[function_component(MyComponennt)]
+/// fn my_component(props: &Props) -> Html {
+///     let greeting = props.callback.emit("Yew".to_string());
+///
+///     html! {
+///         <>{ &greeting }</>
+///     }
+/// }
+///
+/// #[function_component(UseCallback)]
+/// fn callback() -> Html {
+///     let counter = use_state(|| 0);
+///     let onclick = {
+///         let counter = counter.clone();
+///         Callback::from(move |_| counter.set(*counter + 1))
+///     };
+///
+///     // This callback depends on (), so it's created only once, then MyComponennt
+///     // will be rendered only once even when you click the button mutiple times.
+///     let callback = use_callback(
+///         move |name| format!("Hello, {}!", name),
+///         ()
+///     );
+///
+///     // It can also be used for events.
+///     let oncallback = {
+///         let counter = counter.clone();
+///         use_callback(
+///             move |_e| (),
+///             counter
+///         )
+///     };
+///
+///     html! {
+///         <div>
+///             <button {onclick}>{ "Increment value" }</button>
+///             <button onclick={oncallback}>{ "Callback" }</button>
+///             <p>
+///                 <b>{ "Current value: " }</b>
+///                 { *counter }
+///             </p>
+///             <MyComponennt {callback} />
+///         </div>
+///     }
+/// }
+/// ```
+#[hook]
+pub fn use_callback<IN, OUT, F, D>(f: F, deps: D) -> Callback<IN, OUT>
+where
+    IN: 'static,
+    OUT: 'static,
+    F: Fn(IN) -> OUT + 'static,
+    D: PartialEq + 'static,
+{
+    let callback = use_state(|| -> RefCell<Option<Callback<IN, OUT>>> { RefCell::new(None) });
```

Comment:
You are correct, updated

Reply:
then I'm wondering whether we need to add use_callback at all? but React uses useCallback a lot for performance.

Reply:
another issue, the array of dependencies is not passed as arguments to the callback, if callback wants to consume deps, that might not be convenient.

Reply:
```rust
let oncallback = {
    let counter = counter.clone();
    let counter2 = counter.clone();
    use_callback(
        move |e| *counter2,
        counter
    )
};
```

vs

```rust
let oncallback = {
    let counter = counter.clone();
    use_memo(
        move |counter| {
            Callback::from(move |e| **counter)
        },
        counter
    )
};
```

Reply:
@futursolo any advice?

Reply:
I believe that React has a separate `useCallback` primarily due to `useMemo(() => () => {}, [])` looks a little bit complicated. I think in this case, if `use_callback` manages `deps` for you, it might still be worth it.

So one can do something like:
```rust
use_callback(move |deps| move |in_type| {}, deps);
// or have some magic by wrapping it in a custom closure.
use_callback(|deps, in_type| {}, deps);
```

Reply:
`use_callback(|deps, in_type| {}, deps);` looks interesting, I'll have a try.